### ISSUE TITLE
Project Planning

### ISSUE THREAD

**Issue by matzf** [scionproto/scion#4348]
In a recent [TC Implementation meeting](https://github.com/scionproto/scion/issues/4327#issuecomment-1573359749), we've discussed the idea to develop and publish roadmaps for the development of the open-source SCION implementation.
The idea is to **make planned development activity visible**, to communicate planned features to users and help coordinate between developers.
It's not meant to be a binding commitment, and contributions outside of this process are still very welcome.

To start, we'll use a very simple process:
- Gather input from community on planned work, desired features etc. in this github issue
- The TC Implementation then compiles a roadmap from this input and publicly announces it

**Everybody**: please comment to this issue on development topics that you're either planning to work on yourself or that you would be keen to have someone else work on.

**Comment by matzf** [comment]
Very broadly, a goal of the SCION Association currently is to make the open-source SCION implementation directly operationally usable; the open-source SCION implementation should be (at least minimally) operationally viable on its own, not "only" serve as basis for an _actually_ usable proprietary implementation.

- Router and SCION-IP gateway performance
  Performance improvements required for router / gateway to become minimally viable alternatives to proprietary implementations
- Critical missing features:
  - peering links
  - configurable path and routing policies in gateway
- Stabilize inter-AS communication protocols
  Use appropriate network protocols between control services of different ASes, path control for inter-AS communication
- Extended beaconing and path policy configuration language
  Required for connecting e.g. SCION education network (strictly non-commercial) to production network
- Overlapping ISDs
  Required in future deployments, should be done as long as we can still do protocol-breaking changes
- Router/control service key management
  Forwarding key and DRKey rollover
- Authenticated SCMP control messages
  Partially implemented. Todo are validation and processing of SCMP messages at end host, rate limiting.
  Pending firm decision on whether or not this feature will be included in SCION or not.
- Support SCION-native applications on larger variety of platforms
  Dispatcher-less application stack (allowing multiple independent SCION applications without OS support or central component)
- General usability enhancements
  - Documentation (reference manuals, guides, tutorials)
  - Software packaging / releases
  - Sanitize configuration file tree and formats
  - Tooling

Myself, and in the near future one more software engineer directly employed by the SCION Association, are prioritizing these topics.

**Comment by johnstudarus** [comment]
Is there any need for CI automation and regression testing? (Maybe that falls under Software packaging/releases). Or continuous validation of the SCION stack working on various bare metal service providers? @alicecuii - perhaps this would be a good project for you? Or if you want to tackle something larger, I'd love to see a message bus ported over to native SCION.

@JbainMartincoit - I'd like you to stay in the loop and see how you can help out with "General usability enhancements
Documentation (reference manuals, guides, tutorials)"

**Comment by matzf** [comment]
Looking at [a recent CI run](https://buildkite.com/scionproto/scion-nightly/builds/1721) gives a bit of an overview of the set of checks that we run. This gives us a relative broad coverage already. Of course there can always be more and better tests, but I don't know of a subsystem or topic that is simply lacking tests.

One specific topic related automated testing where we could use help is this: there is a test harness called `braccept` (for *b*order *r*outer *accept*ance tests). This sets up a SCION border router in a specific configuration, sends individual packets to it and validates that the router behaves as expected. Currently, this is very strict and tolerates no deviations from the open-source "reference implementation". Ideally, we'd want to reuse the same test harness to assess other router implementations, but for this it would need to be redesigned so it can accept and report on different behavior allowed by the of the specification (with all its "SHOULD"s and "MAY"s etc). The data plane specification document is currently being worked on, so it could be a good timing to work on this.
If this sounds like an interesting topic, we could create a github issue describing the requirements, so that you could then start to work out the details.

**Comment by JordiSubira** [comment]
Currently, I am working on the following items on that list:

- Authenticated SCMP control messages
- - Partially implemented here https://github.com/scionproto/scion/pull/4255
- - Accepted change in the specification here https://github.com/scionproto/scion/pull/4300
- - (Ongoing) Implement changes in latter point
- Dispatcher-less application stack (allowing multiple independent SCION applications without OS support or central component)
- - (Ongoing) Initial effort here: https://github.com/scionproto/scion/pull/4344.

I also have interest in tackling:
- Router/control service key management

I would kindly ask for whoever is interested on this to coordinate together :)

**Comment by benthor** [comment]
_There now is a [document](https://github.com/netsys-lab/organizational-things/blob/main/ongoing_work.md) with ongoing work related to SCION at OVGU. For convenience, I've pasted the current status below:_

### Ongoing Work in NetSysLab - Relevant to SCION Association Roadmap

- Make the Go reference border router's data plane exchangeable to allow using hardware implementations like the Tofino or XDP from the Go code (Lars + Robin)
- BPF implementation of SCION Border Router (Lars)
- Implement AF_XDP as interface between XDP router and Go control plane (Lars + Robin)
- [ID-INT](https://github.com/netsys-lab/id-int-spec) development (Lars)
  - ID-INT support in Go BR
  - Host ID-INT
- [Support for Peering Links](https://github.com/scionproto/scion/pull/4299) (Thorben / waiting on upstream)
- [SCION Certificate Authority](https://github.com/netsys-lab/scion-ca) and [renew service for endhosts](https://github.com/netsys-lab/scionlab-cert-renewer) (Marten)

### Further Ongoing Work - Related to SCION

- Extend SEED with Tofino support (Lars + Robin)
- [BitTorrent over SCION](https://github.com/netsys-lab/bittorrent-over-scion) improvements and GUI (Marten / student project) 
- DNS-over-QUIC support for SCION (Thorben / student project)
- IPFS over SCION, SCION support in libp2p (Marten / student project)
- [Hercules](https://github.com/netsec-ethz/hercules) ongoing work and improvements (Marten)
- Deadline-aware Multipath Transport Protocol (Tony)
- [Scriptable path selector](https://github.com/netsys-lab/pan-lua) (Thorben)
- Evolutionary Multi-objective optimization for optimal path selection (Tony / student project)
- Bitcoin over SCION - Resilience against partitioning and other routing attacks (Tony / student project)
- Splitting the border router's AES-CMAC validation and forwarding mechanisms to allow more flexible implementation in hardware (Lars)
  - Implement validation in NetFPGA (Björn + Florian)
  - Implement forwarding in P4 on Tofino switch
    - With support for external accelerator and ID-INT (Robin)
    - With support for EPIC (David G.)
- [PAN-bindings](https://github.com/lschulz/pan-bindings) for C, C++ and Python (Lars)

### Future Plans

- AES-CMAC Linux kernel implementation
- Implement beacon filtering

**Comment by matzf** [comment]
The SCION Association's TC Implementation has compiled a roadmap document.
Find the full document here: [2023-10 Open-Source SCION Implementation Roadmap](https://www.scion.org/assets/docs/2023-10%20Open%20Source%20SCION%20Implementation%20Roadmap.pdf)

*Summary*

#### Goals and Ambition
The open-source implementation of SCION aspirs to be **complete**, i.e. not missing any required components or critical features, and and of sufficient quality (in terms of correctness, performance and usability), and with sufficient documentation to be minimally **operationally viable**.

#### Scope
- Infrastructure components for a SCION network, namely the SCION control service and router, and tooling for the SCION control plane PKI
- End host SCION network stack, tools, and application programming libraries to build SCION native applications
- SCION-IP-Gateway

The scope of the project and these components is the “base” SCION protocol. For the time being expressly **not in scope** are experimental SCION-extensions such as Colibri, or EPIC.

#### Releases
When major work topics are complete, we'll publish releases consisting of:
- Source Code: git version tag and source code bundles
- Program binaries and installation packages
- Documentation

#### Development focus

The focus for the planned work in this period is to improve operational readiness and deployability of the open-source components.

##### Planned work topics:

Infrastructure components (router, control service, gateway):
* Peering links
* Router performance improvements 
* Stabilize control plane communication protocols 
* Extend control service beaconing policy configuration
* Rolling key change for hop keys in control service and router
* SCION-IP-Gateway performance improvements
* SCION-IP-Gateway configurable path and routing policies
* Documentation
* Software Packaging
* General configuration file structure overhaul
* Compatibility assessment tool for router

End host stack, SCION-native applications:
* Adopt end-host autoconfiguration into official end-host stack
* Dispatcher-less application stack

##### Possible timeline

Possible sequence of work items and rough estimated time of arrival, covering the period of about one year starting from now.

<sup>*Disclaimer*: this is based on rough estimates for the effort and many of these topics will require upfront discussion and detailed design work. The timing may depend on the availability of independent collaborators and their voluntary contributions. The included release milestones illustrate that we will release after completing major work topics and should not be mistaken to indicate a commitment to release these version numbers at the indicated dates or feature sets.</sup>

![image](https://github.com/scionproto/scion/assets/42933359/652601f0-81c0-4e56-96a3-00b0fb58eaf2)

-----

We'll update this post with links to issues/milestones to track the progress of individual topics.

This roadmap intends to guide the development work for the coming year or so, but it's not meant to be set in stone. We may amend the roadmap if circumstances change, and everyone is still welcome to contribute to this planning process by commenting  or discussing in the chat.

**Comment by jiceatscion** [comment]
https://docs.google.com/document/d/e/2PACX-1vS7-FzHRZ9Utz6cO6TY4IDZ_54a32euX5sWm8OZc2xV51tggU6mIkSbMDHn02cBdBSIsSEGuOrjWHwX/pub
